# Introduction

This pipeline will call the SNPs and Indels for WES sequencing or Panel sequencing 

### **1. Requirements**

**Python** 3.8

**GATK** v4.1.9.0

**FastQC** v0.11.9

**Trimmomatic** 0.39

**Bwa** 0.7.17-r1188

**Samtools** 1.11

**Bedtools** v2.29.2

All of those could be installed by `conda`

### **2. Input and configuration:**

- The fastq path with fastq file

- The sample name

- The output path

- the config file 

```
[database]
ref = /nfs/databases/database_NGS/hg19/reference/ucsc.hg19.fasta
refdict = /nfs/databases/database_NGS/hg19/reference/ucsc.hg19.dict
genomefile = /nfs/databases/database_NGS/hg19/reference/ucsc.hg19.genome.txt
hapmap = /nfs/databases/database_NGS/hg19/gatk_bundle/hapmap_3.3.hg19.sites.vcf
omni = /nfs/databases/database_NGS/hg19/gatk_bundle/1000G_omni2.5.hg19.sites.vcf
db1000g_phase1_snp = /nfs/databases/database_NGS/hg19/gatk_bundle/1000G_phase1.snps.high_confidence.hg19.sites.vcf
dbsnp = /nfs/databases/database_NGS/hg19/gatk_bundle/dbsnp_138.hg19.vcf
mills_indel = /nfs/databases/database_NGS/hg19/gatk_bundle/Mills_and_1000G_gold_standard.indels.hg19.sites.vcf
bedfile = /nfs/databases/database_NGS/hg19/bedfile/tEXOME2.bed
targetbed = /nfs/databases/database_NGS/Agilent/S07604514_Regions.bed
adapter = /nfs/databases/database_NGS/Trimmomatic-0.39/adapters/TruSeq3-PE-2.fa

[parameter]
cpu = 10
```

`ref=` is the path of reference genome (hg19 or hg38)

`refdict=` is the path of reference genome dict file, could be generated by 

```
gatk CreateSequenceDictionary -R ucsc.hg19.fasta -O ucsc.hg19.dict
```

`genomefile=` is the path of genome file,  could be generated by 

```
less ucsc.hg19.dict | grep ^"@SQ" | awk -F '[\t:]' '{print $3"\t"$5}' >  ucsc.hg19.genome.txt
```

`hapmap=` and`omni=`and `db1000g_phase1_snp=`and `dbsnp=`and `mills_indel=` is the corresponding gatk bundle file of the different genome

`bedfile =` is the path of bedfile for calling variant

`targetbed=` is the path of bedfile for capturing region

`adapter=` is the path of adapter information for raw fastq 

### **3. Usage example**

python wes_gatk.py  \

--fqfolder /data_obs/NGS_Data/201024_A01200_0048_BHTMKJDMXX/tZTJCYCBhb-NP22FM0398/FASTQ/ \

--sample tZTJCYCBhb-NP22FM0398 \

--output /nfs/pipeline_validate/tZTJCYCBhb-NP22FM0398 \

--config /home/taojingfen/projects/20200927_WES/WES_GATK/scripts/wes_config.ini \
